### PR TITLE
Autogenerate c_cpp_properties.json

### LINKED ISSUE #438: The Extension should automagically fill the c_cpp_properties.json, so intellisense works out of the box
EDIT: 07/03/2018: My understanding of the issue has improved since I opened it. I think that the extension should fill the intellisense config file on the fly based on the board being selected by the user.

To my knowledge, the Arduino IDE checks for libraries in 4 places: the main Arduino installation path (in windows c:\program files.\...), any board-specific version of the libraries for this board ("...\appdata\local\..."), your sketchbook location and a "libraries" folder for that particular sketch. Ideally, you would install your libraries in your sketchbook location, so they are available for all your projects.

I've just installed VS Code and the Arduino plugin, and it seems to be able to know that my sketchbook location is not the default, as I can see the examples of the libraries that I've installed there, and I can compile my sketch. In any case, I get tens of intellisense errors, as intellisense seems incapable to follow the paths for the includes. I can add the sketchbook location libraries to the c_cpp_properties.json, and either add all the libraries used there (painfull) or add the base "libraries" folder and then tweak the includes to have partial paths (less painfull), but even in these cases, intellisense will complain that it can't find any of the libraries this files might use. And I would have to do it for all my projects, one by one.

What am I doing wrong? why doesn't this plugin pick the usual library locations? Is it normal to have 100s of intellisense errors? is that really the way it's meant to be?

Sorry, I'm new to Visual Studio and Visual Studio Code, and I'm quite sure that the problem it's me not being able to find out the logic behind them, but I can't find any info on the internet either.

I'm using windows 10 64 bits, and my arduino Ide is at the latest version. The relevant part of my  c_cpp_settings.json of this project is:

```{
            "name": "Win32",
            "includePath": [
                "C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.11.25503/include/*",
                "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/um",
                "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/ucrt",
                "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/shared",
                "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/winrt",
                "${workspaceRoot}",
                "C:\\Users\\Joel\\AppData\\Local\\Arduino15\\packages\\SODAQ\\hardware\\samd\\1.6.14\\cores\\arduino"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE"
            ],
            "intelliSenseMode": "msvc-x64",
            "browse": {
                "path": [
                    "C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.11.25503/include/*",
                    "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/um",
                    "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/ucrt",
                    "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/shared",
                    "C:/Program Files (x86)/Windows Kits/10/Include/10.0.16299.0/winrt",
                    "${workspaceRoot}",
                    "C:\\Users\\Joel\\AppData\\Local\\Arduino15\\packages\\SODAQ\\hardware\\samd\\1.6.14\\cores\\arduino"
                ],
                "limitSymbolsToIncludedHeaders": false,
                "databaseFilename": ""
            }
        }
```

This is a screenshot of my project

### LINKED ISSUE #808: identifier "Serial" is undefined
My VS Code have a problem:identifier "Serial" is undefined
but I upload program to my arduino nano OK.
this is my arduino file:
``` c
void setup(){
    pinMode(LED_BUILTIN, OUTPUT);
    Serial.begin(9600);
}

void loop(){
    digitalWrite(LED_BUILTIN, HIGH);
    Serial.println("hello arduino 2333333333333");
    delay(1000);
    digitalWrite(LED_BUILTIN, LOW);
    delay(1000);
}
```

this is arduino.json
``` json
{
    "board": "arduino:avr:nano",
    "configuration": "cpu=atmega328old",
    "programmer": "AVRISP mkII",
    "sketch": "serialPort.ino",
    "port": "COM9"
}
```

this is c_cpp_properties.json
```  json
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "C:\\\\Arduino\\\\tools\\\\**",
                "C:\\\\Arduino\\\\hardware\\\\arduino\\\\avr\\\\**",
                "C:\\\\Arduino\\\\hardware\\\\arduino\\\\avr\\\\cores\\\\arduino\\\\**",
                "C:\\\\Arduino\\\\hardware\\\\tools\\\\avr\\\\avr\\\\include\\\\avr\\\\**",
                "C:\\\\Arduino\\\\hardware\\\tools\\\\**", 
                "C:\\\\Arduino\\\\hardware\\\tools\\\\avr\\\\avr\\\\include\\\\**"
            ],
            "forcedInclude": [
                "C:\\\\Arduino\\\\hardware\\\\arduino\\\\avr\\\\cores\\\\arduino\\\\Arduino.h"
            ],
            "intelliSenseMode": "gcc-x64",
            "compilerPath": "C:/mingw64/bin/gcc.exe",
            "cStandard": "c11",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```

### LINKED ISSUE #966: Express server is started on random port when vscode-arduino is activated
## VSCode info
Version: 1.41.1
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-20T11:11:01.597Z
Electron: 6.1.7
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Linux x64 5.5.2-arch2-2

## Extension version
0.2.29

## Description:
A random port is opened on the machine on which the extension is running, with an Express server listening. The opened port is available form other hosts on the same network.

## Steps to reproduce (linux host) :
1. install extension
2. run `sudo netstat -tulpan | grep LISTEN` and note the port (here 44873)
````
tcp6       0      0 :::44873                :::*                    LISTEN      16511/electron      
````
3. run `curl localhost:44873` (replacing 44873 with port noted in 2.
````
<html>
  <head>
    <style title="themestyle"> </style>
    <script type="text/javascript">
      // inherit theme styles from vscode ide
      var decode = function (s) {
        return decodeURIComponent(s.replace(/\+/g, " ")); 
      };
      var getStyleSheet = function (unique_title) {
        for(var i=0; i<document.styleSheets.length; i++) {
          var sheet = document.styleSheets[i];
          if(sheet.title == unique_title) {
            return sheet;
          }
        }
      };
      var match,
          search = /([^&=]+)=?([^&]*)/g,
          query  = window.location.search.substring(1),
          queryParams = {};
      while (match = search.exec(query)) {
        queryParams[decode(match[1])] = decode(match[2]);
      }
      var supportedThemes = ["vscode-dark", "vscode-light"];
      var theme = "vscode-dark"; // default theme.
      if (supportedThemes.indexOf(queryParams["theme"]) > -1) {
        theme = queryParams["theme"];
      }
      var bgcolor = queryParams["backgroundcolor"];
      var color = queryParams["color"];
      var stylesheet = getStyleSheet("themestyle");
      stylesheet.insertRule(".vscode-theme { background-color: " + bgcolor + "; color: " + color + "; }", 0);
      stylesheet.insertRule(".vscode-theme .theme-bgcolor { background-color: " + bgcolor + "; }", 1);
      stylesheet.insertRule(".vscode-theme .theme-color { color: " + color + "; }", 2);
      stylesheet.insertRule("html > body { background-color: " + bgcolor + "; color: " + color + "; }", 3);
      document.documentElement.className = theme + " vscode-theme";
    </script>
  <link href="styles/app.css" rel="stylesheet"></head>
  <body>
    <div id="mainContent"></div>
  <script type="text/javascript" src="app.js"></script></body>
</html>
````
4. disable extension
5. run `sudo netstat -tulpan | grep LISTEN` again
> nothing

## Expected behavior

No port is opened, or at least opened on a local only mode.

Culprit seems to be this file : https://github.com/microsoft/vscode-arduino/blob/master/src/arduino/localWebServer.ts

### LINKED ISSUE #969: INO Files defines undefined but can jump to definition
Within Arduino INO files, values that are defined in project level header files make intellisense throw the following error:
image

However, as shown, it appears it is able to jump to definition of the define. Project compilation also works fine without an issue

### LINKED ISSUE #1139: changing the default c_cpp_properties created
when i start working with it a default c_cpp_properties is created. this includes:
```
"includePath": [
                "C:\\Program Files (x86)\\Arduino\\tools\\**",
                "C:\\Program Files (x86)\\Arduino\\hardware\\arduino\\avr\\**"
            ],
```
however for intellisense to work correctly i also need to add `"C:\\Program Files (x86)\\Arduino\\hardware\\tools\\avr\\avr\\**"`
is there a way to also add this line to the "default creation" of the file or do i need to add it manually everytime i start a new project without adding it to c_cpp.default which would include it in every c/c++ project?

### PR DESCRIPTION
**Pull request by iFreilicht**
Fixes #438 
Fixes #808 
Fixes #969 
Fixes #1139 
Fixes #966

Potentially fixes other issues that stem from an incorrectly set up c_cpp_properties.json.

This PR adds an autogeneration mechanism to the extension. This way, library paths and macro definitions are added to c_cpp_properties.json automatically.

See the original PR #1107 for more info. It was rebased by @hlovdal

### CODE REVIEW

**Review by adiazulay — CHANGES_REQUESTED** [review]
I took a quick look at the PR and overall I like the approach to the problem. This worked well when generating a new c_cpp_properties.json file for the Uno and an EPS32 board, but there were a few issues I came across listed below.

1.  c_cpp_properties.json fails to generate if there is a already a version of the file generated by and earlier version of the extension
2. c_cpp_properties.json fails to generate if there are additional header and cpp files in the workspace
3. c_cpp_properties.json is not generated when running tests
4. IntelliSense doesn't work for Arduino Nano 33 BLE Sense (Mbed OS), unable to find "PinNames.h"

**Comment by iFreilicht** [comment]
@hlovdal could you add me as a collaborator to your repo? I'll try and address the issues raised.

1. Should be a fixable bug, I'll look into that
2. You mean if there are files that are or are not included? I was able to work with relatively complex sketches that include a few headers from the same workspace, and the c_cpp_properties.json was generated perfectly.
3. Ah, that's an oversight, should be easy to fix
4. That is quite weird, pins are defined in Arduino in "pins_arduino.h", not "PinNames.h". Is this a board you installed from an additional board manager URL?

**Comment by hlovdal** [comment]
Collaboratorship granted.

**Comment by adiazulay** [comment]
@iFreilicht 

2. It was for an included file, I'll get better repro steps after next week.
4. I'm pretty sure I installed this board through the board manager in the Arduino IDE, I can try to dig into it a bit more and see if I can figure out which include path is missing.

**Review by gugavaro — APPROVED** [review]

File: src/arduino/intellisense.ts
Comment (fearthecowboy) [reply src/arduino/intellisense.ts:94]: Really? I recall having issues trying to use C++17 with arduino a year or so ago. Compilation worked, but the code was trash.

Has C++ 17 support been added to Arduino in the last year or so?

File: src/arduino/arduinoContentProvider.ts
Comment (fearthecowboy) [inline src/arduino/arduinoContentProvider.ts:50]: FYI - you can just `return this._webserver.start();` and return the promise (and then you don't have to await here ) and you don't have to mark this method `async`

File: src/arduino/localWebServer.ts
Comment (fearthecowboy) [inline src/arduino/localWebServer.ts:39]: again, this really shouldn't be `async` -- if you're returning a promise without any awaits, just return the promise.

File: BRANCHNOTES.md
Comment (fearthecowboy) [inline BRANCHNOTES.md:434]: O_o

I'm a bit confused -- we're removeing addLibPath from the commands (see `package.json `), but it's still used internally a bunch, so we're not removing the implementation?

Does it's use internally (and in the UI ) still work?

(I'm just not really sure if more code should be removed, or what...)

**Review by fearthecowboy — COMMENTED** [review]
I'm generally OK with the changes

I'm assuming this was inherited with the rather thin unit tests here. Not a lot we can do, but I don't know how we're verifying actual functionality here.

**Comment by adiazulay** [comment]
> @iFreilicht
> 
> 2. It was for an included file, I'll get better repro steps after next week.
> 3. I'm pretty sure I installed this board through the board manager in the Arduino IDE, I can try to dig into it a bit more and see if I can figure out which include path is missing.

2. The issue appears to stem from Arduino failing to verify the sketch while there is a parent folder/main sketch name mismatch. I don't think this is an issue with this PR.

3. This looks like how the mbed arduino core is organized. The pin names come from the target folder and are specific to different MCUs.
I fixed it by adding the core folder recursively, is there any reason not to do this by default? 

```
"C:\\Users\\adazulay\\AppData\\Local\\Arduino15\\packages\\arduino\\hardware\\mbed\\1.1.6\\cores\\arduino\\**",
```

File: BRANCHNOTES.md
Comment (iFreilicht) [reply BRANCHNOTES.md:434]: Ah, that's an oversight. Calling the `addLibPath` command from the UI doesn't make sense anymore, as the library paths will be added automatically, but `addLibPath` is still called vie the "Include Library" button in the Library Manager (see `LibraryItemView.tsx`).

So `addLibPath` should be renamed to `includeLibrary` as that's all it does now. As the library path will be detected and automatically added afterwards, there's no need for the "old" addLibPath anymore.

**Comment by iFreilicht** [comment]
> I'm generally OK with the changes
> 
> I'm assuming this was inherited with the rather thin unit tests here. Not a lot we can do, but I don't know how we're verifying actual functionality here.

So I and a bunch of other people have tested this quite a bit, you can see our conversation in this [gitter chatroom](https://gitter.im/vscode-arduino-ew-fork/community?utm_source=share-link&utm_medium=link&utm_campaign=share-link). Actually, while looking at it, I found another issue regarding `addLibPath` I raised months ago that wasn't fixed yet, so I'm gonna do that as well.

**Comment by iFreilicht** [comment]
> @adiazulay 
>
>     1. The issue appears to stem from Arduino failing to verify the sketch while there is a parent folder/main sketch name mismatch. I don't think this is an issue with this PR.

I've actually reported this before #386. In some cases, vscode-arduino can handle this, in some cases it can't, but the Arduino IDE itself is very clear that a sketch has to be in a folder matching its name.

>     2. This looks like how the mbed arduino core is organized. The pin names come from the target folder and are specific to different MCUs.
>        I fixed it by adding the core folder recursively, is there any reason not to do this by default?
> 
> 
> ```
> "C:\\Users\\adazulay\\AppData\\Local\\Arduino15\\packages\\arduino\\hardware\\mbed\\1.1.6\\cores\\arduino\\**",
> ```

Hmm, right now the extension is very specific about the folders it includes, so it will always only choose the direct parent folder of header files that need to be included.
This makes the `c_cpp_properties.json` longer and more verbose, but that may not be a bad thing.

One thing I noticed is that there's actually **42 different files** called `PinNames.h` inside `mbed/1.1.6/cores/arduino`. So I'm not sure how the compiler knows which one to include.

File: src/arduino/intellisense.ts
Comment (adiazulay) [reply src/arduino/intellisense.ts:94]: you're probably right about that, C++17 is what the extension currently defaults to, but I can't find any documentation on what the arduino complier uses

**Comment by adiazulay** [comment]
@iFreilicht I dug in a bit and found where the target is being selected.

1. The "Fully Qualified Board Name" indicates which board is used. -fqbn compiler argument
2. Board name is then compared to boards.txt (C:\Users\user\AppData\Local\Arduino15\packages\arduino\hardware\mbed\1.1.4\boards.txt)
`nano33ble.build.variant=ARDUINO_NANO33BLE`
3. in the variant folder includes.txt points to the correct target (C:\Users\user\AppData\Local\Arduino15\packages\arduino\hardware\mbed\1.1.4\variants\ARDUINO_NANO33BLE)
`-iwithprefixbefore/mbed/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/TARGET_ARDUINO_NANO33BLE`

File: src/arduino/intellisense.ts
Comment (fearthecowboy) [reply src/arduino/intellisense.ts:94]: We should probably find out. I'm 60% sure arduino uses(defaults?) to C++11 :smile: 

IIRC, there were a lot of things around lambdas/captures that were flakey as heck, but I can't for the life of me remember much past that.

Comment (iFreilicht) [reply src/arduino/intellisense.ts:94]: So, I installed the current Arduino IDE 1.8.13 on Windows, and it is bundled with avr-gcc 4.8.1, which was originally released in 2013. Looking at the [online docs of the bundled libstd](https://gcc.gnu.org/onlinedocs/gcc-4.8.1/libstdc++/manual/), only C++11 is mentioned in name. I think it's safe to say this version of gcc doesn't support C++14, let alone C++17.

In theory, it seems to be possible to replace this old version of gcc with newer ones in the toolchain, I've found some threads were people said they saw someone else do it before, but I think it's safe to say that this won't be the default use-case, so C++11 is the right value here.

**Comment by hlovdal** [comment]
Just a heads up, I am working on rebasing again on top of the latest `develop`, I expect to be finished in a couple of days.

**Comment by iFreilicht** [comment]
Thanks! I'll hold off on pushing anything to your branch for now. Just tell me when it's ready.

**Comment by adiazulay** [comment]
@hlovdal I'm getting ready to merge #1017 today, would be worth waiting until this afternoon to start working through this rebase.

**Comment by adiazulay** [comment]
@hlovdal @iFreilicht I merged ```develop``` into this pr, but I don't have push access. Everything looks like it's working right with the CLI and is ready merge unless you have any pending commits.

**Comment by iFreilicht** [comment]
I'd still like to fix those two small annotations you had. @hlovdal, can you give Adi collaborator rights as well, so he can push the new rebase? I'll do the fixes then and after that we can merge :+1:

**Comment by adiazulay** [comment]
I also opened a PR on the fork (https://github.com/hlovdal/vscode-arduino/pull/1) , so you can just merge that too.

**Comment by hlovdal** [comment]
Please do not merge anything into develop yet.

In the previous rebase I did I made significant effort into making the rebased branch as identical as possible to the original branch, only with modifications like updating the commit reference in the commit message to 5fd7ef80 for instance, since the branch was moving forward quite a distance and I did not want any changes other than that.

In the rebase I am currently working on I have fixed the branch so that every single commit compile, lint and test correctly. This is required for bisect to work (and a good thing regardless). I have also done some general version history clean up like for instance one commit added

```typescript
    public port: StrSetting = new StrSetting();
    public board: StrSetting = new StrSetting();
    public sketch: StrSetting = new StrSetting();
    public output: StrSetting = new StrSetting();
    ...
```

which later were changed to

```typescript
    public port = new StrSetting();
    public board = new StrSetting();
    public sketch = new StrSetting();
    public output = new StrSetting();
    ...
```

so I modified the original commit to just use the latter version from the start. Both are identical from the compilers's point of view, but now the version history contains less back and forth noise.

So the final result for old and new rebase should be virtual identical, but with some difference in the commits the result is composed from.

The current obstacle I am working on is the update made by pull request 1017 which merged the `update` and `uploadUsingProgrammer` functions into one combined function with some changes while this branch has merged the `update`, `uploadUsingProgrammer` and `verify` functions into one combined function with some significant changes compared to the mainline branch. This very much does not automatically resolve itself by tools (and in practice not manual effort either). The solution is of course to split up in smaller commits, and I think I am almost done with that. So I expect to be finished soon, if it had not been for 1017 I would have been done earlier this week.

**Comment by hlovdal** [comment]
@adiazulay I have added you as collaborator. Feel free to make and push any changes and modifications to the branch (but do not merge to develop). You and @iFreilicht do not have to hold back new commits or modifications until I am done with the rebase, I will have no problem with picking up and adopting changes you push in the meantime.

(Althoug preferably do not rebase content older than commit e12aa697 (Updated log, ...))

**Comment by adiazulay** [comment]
@hlovdal  I merged instead of rebased the develop branch and resolved the issues with upload/upload with programmer/verify.
I'm going to squash this branch when I merge it, so history in develop is going to stay clean.

**Comment by adiazulay** [comment]
@iFreilicht my PR is merged, you should be able to push your changes now.
@hlovdal let me know if you still want to continue rebasing the branch before merge.

**Comment by hlovdal** [comment]
Yes I am still working on it but have been busy with other things the last couple of days. I managed to finish the imerge rebase operation in the weekend, so the major work and challenge is done. There are a few commits on near the top of the branch that does not compile and a few other things I need to clean up, but it is getting close.

**Comment by adiazulay** [comment]
@iFreilicht do you think you'll be able to get the last couple of fixes in this week? I'm hoping to merge this by Friday.

Let me know if there anything I can do to help.

**Comment by hlovdal** [comment]
@adiazulay Friday is too soon, neither your nor my branch is quite ready yet.

Some update from me: before Christmas I was busy working with a couple of woodworking gift projects, I was away for several days and after returning I had some problem with the pc not unencrypting the disk on startup after an update (I had to change a luks.uuid value). But now my pc is working and I am back in business and have finished my rebased branch to a state where it can be reviewed.

The result of my rebased branch is very similar to your branch where you merged and made additional changes in one commit, 
however there are some errors and issues with your merge branch when comparing with my rebased brach:

----

The change from
`        "programmer=" + selectProgrammer`
to
`        "programmer=arduino:" + selectProgrammer`
has been discarded/lost.

----

```C
-        if (!await this.runPrePostBuildCommand(dc, env, "pre")) {
+        if (dc.prebuild && !await this.runPrePostBuildCommand(dc, env, "pre")) {
             return false;
         }
```

This is the wrong place to test the value of `dc.prebuild`. This is already tested inside `runPrePostBuildCommand` with the existing `cmdline` test (which ought to be inverted) which is the right place to test this. However the merge commit cb3b978d5 failed to put the last return true statement at the correct place at the end (like it was originally when Uli extended `runPreBuildCommand` into `runPrePostBuildCommand` in commit 209c9b7), so it returns undefined instead of true when `dc.prebuild` is falsy. I guess that might have been the motivation for modifying the test above to compensate for that?

----

`CliUpload` and `CliUploadProgrammer` has been added in `src/arduino.ts`, however the `build` function is never called with those values. I assume this is an oversight and that those should have been used in `src/extension.ts`?

----

A couple of "--porgrammer" instead of "--programmer".

----

Some spelling mistakes (including pre-existing) of "available".

----

To make comparing the two branches simpler I have created a branch `intellisense-autoconfig.rebased.compare` on top of `intellisense-autoconfig.rebased` with a few formatting/rename changes to make the branches more similar.

When starting on this second rebase I used the merge commit cb3b978d as reference, and to avoid a moving target I kept working towards that, so some of the changes added after that is missing but I am not sure about all of them and instead of guessing on if some of the things that is different is significant I would like to have you compare the branches and provide some feedback on the things that are different.

I have pushed the second rebase branch as `intellisense-autoconfig.rebased2.2021-01-07_001`.

**Comment by adiazulay** [comment]
@hlovdal thanks for the update! I'll review your branch today or tomorrow. Friday was just a goal, but it's no problem waiting a bit longer until everything is ready.

Thanks for all your work on this.

**Comment by adiazulay** [comment]
@hlovdal I looked through your branch today. Looks good, much neater than mine. I'll get working on integrating the commits after cb3b978. 

> The change from
> ` "programmer=" + selectProgrammer`
> to
> ` "programmer=arduino:" + selectProgrammer`
> has been discarded/lost.

We've made updates to how programmers are being handled, so  instead of hardcoding ```arduino:``` it gets the vendor associated with the programmer. I'll make sure those changes work with the new branch.

You're right about ```dc.prebuild``` it was able to run no problem with the IDE, but would just hang with the CLI. I'll test your branch to make sure it works.

That's just an oversight on those commands

I'll check all the spelling again, thanks for catching those!

**Comment by adiazulay** [comment]
@hlovdal I opened a new PR with your rebased branch (#1183). I pushed a couple of fixes for the using the CLI.

All your changes look good to me. I'll wait to hear from you before merging the new PR.

Thanks for all the hard work everyone!

**Pull request closed without merging**
